pkg/scheduler/framework/plugins/dynamicresources/allocateddevices.go

```diff
@@ -162,14 +224,41 @@ func (a *allocatedDevices) removeDevices(claim *resourceapi.ResourceClaim) {
 	// Locking of the mutex gets minimized by pre-computing what needs to be done
 	// without holding the lock.
 	deviceIDs := make([]structured.DeviceID, 0, 20)
-	foreachAllocatedDevice(claim, func(deviceID structured.DeviceID) {
-		a.logger.V(6).Info("Observed device deallocation", "device", deviceID, "claim", klog.KObj(claim))
-		deviceIDs = append(deviceIDs, deviceID)
-	})
-
+	var shareIDs []structured.SharedDeviceID
+	var deviceCapacities []structured.DeviceConsumedCapacity
+	if a.enabledConsumableCapacity {
+		shareIDs = make([]structured.SharedDeviceID, 0, 20)
+		deviceCapacities = make([]structured.DeviceConsumedCapacity, 0, 20)
+	}
+	foreachAllocatedDevice(claim,
+		func(deviceID structured.DeviceID) {
+			a.logger.V(6).Info("Observed device deallocation", "device", deviceID, "claim", klog.KObj(claim))
+			deviceIDs = append(deviceIDs, deviceID)
+		},
+		a.enabledConsumableCapacity,
+		func(sharedDeviceID structured.SharedDeviceID) {
+			a.logger.V(6).Info("Observed shared device deallocation", "shared device", sharedDeviceID, "claim", klog.KObj(claim))
+			shareIDs = append(shareIDs, sharedDeviceID)
+		},
+		func(capacity structured.DeviceConsumedCapacity) {
+			a.logger.V(6).Info("Observed consumed capacity release", "device id", capacity.DeviceID, "consumed capacity", capacity.ConsumedCapacity, "claim", klog.KObj(claim))
+			deviceCapacities = append(deviceCapacities, capacity)
+		})
 	a.mutex.Lock()
 	defer a.mutex.Unlock()
 	for _, deviceID := range deviceIDs {
 		a.ids.Delete(deviceID)
 	}
+	for _, shareID := range shareIDs {
+		a.shareIDs.Delete(shareID)
+	}
+	for _, capacity := range deviceCapacities {
+		deviceID := capacity.DeviceID
+		if _, found := a.capacities[deviceID]; found {
+			a.capacities[deviceID].Sub(capacity.ConsumedCapacity)
+			if a.capacities[deviceID].Empty() {
+				a.capacities.Remove(capacity)
+			}
+		}
```

Comment:
Shouldn't it be just like this?
```suggestion
		a.capacities.Remove(capacity)
```